### PR TITLE
[autoscaler] Remove deprecated fields from schema

### LINKED ISSUE #20837: Remove deprecated autoscaler fields [Bug][autoscaler] Retire deprecated autoscaler fields
Deprecated fields like "initial_workers, target_utilization_fraction, autoscaling mode, worker_nodes, head_node` need to be officially retired at this point (we should raise an error if these fields are present).
Anecdotally, some of these fields no longer work well. There's not much interest maintaining them.

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This change cuts off support for deprecated schema fields. It intentionally breaks backwards compatibility with old configs which set a global `min_workers`, use `head_node` or `worker_nodes`, `autoscaling_mode`, `initial_workers`, `target_utilization_fraction`, and `default_worker_node_type` fields. 

## Related issue number

Strictly speaking, this should close https://github.com/ray-project/ray/issues/20837 but there will likely be a follow up PR to clean up some of the now dead code. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wuisawesome** [comment]
A few remaining things to do before merging (I'll remove the do-not-merge flag after doing these).

* Run example minimal, installing the wheel from this CI branch on the head node
* Write the migration doc for these backwards compatibility breaking changes

File: python/ray/tests/test_autoscaler_yaml.py
Comment (wuisawesome) [inline python/ray/tests/test_autoscaler_yaml.py:304]: Note that this test *will* break on this PR because we're intentionally breaking backwards compatibility.

**Comment by wuisawesome** [comment]
cc @scv119 This should be the final PR that needs to make it into 2.0 (we should be able to get it in before the branch cut deadline, but maybe it will be the first cherry-pick).

File: python/ray/tune/examples/tune-default.yaml
Comment (wuisawesome) [inline python/ray/tune/examples/tune-default.yaml:7]: missed this file, at least we know the schema change + unit tests work

File: python/ray/autoscaler/ray-schema.json
Comment (scv119) [inline python/ray/autoscaler/ray-schema.json:309]: hmm looks this hasn't been deprecated previously?

**Comment by scv119** [comment]
also if an user is on an old yaml, what's the behavior when they switch over to ray 2.0? Is it possible to provide a tool to upgrade the yaml from the old to the new?

File: python/ray/autoscaler/ray-schema.json
Comment (DmitriGekhtman) [reply python/ray/autoscaler/ray-schema.json:309]: Looks like we didn't mark it as such in the schema, but we have been logging deprecation warnings for this field and worker_nodes for over a year.
https://github.com/ray-project/ray/blame/c7ae787cc8468d8087aa3483b472f40624bd4a22/python/ray/autoscaler/_private/util.py#L189-L190

**Comment by DmitriGekhtman** [comment]
> also if an user is on an old yaml, what's the behavior when they switch over to ray 2.0? 

Validation error, we've been emitting deprecation warnings for a long time.

I think mentioning removed fields in migration notes would also be good.

File: python/ray/autoscaler/_private/commands.py
Comment (DmitriGekhtman) [inline python/ray/autoscaler/_private/commands.py:931]: nice going. that's sharp thinking for past 11pm. 
this is definitely a hack to undo some hacks, but we'll take it

File: python/ray/tests/test_autoscaler.py
Comment (DmitriGekhtman) [inline python/ray/tests/test_autoscaler.py:487]: Somehow I though that the modifications needed for this file would be much worse.
Glad that this is all.

File: python/ray/tests/test_autoscaler_yaml.py
Comment (DmitriGekhtman) [reply python/ray/tests/test_autoscaler_yaml.py:304]: i have no idea what this is, maybe add a docstring

**Comment by DmitriGekhtman** [comment]
migration docs would indeed be good

are we doing a general migration doc for Ray 2.0.0

**Review by DmitriGekhtman — APPROVED** [review]
migration doc good to have

**Comment by DmitriGekhtman** [comment]
> also if an user is on an old yaml, what's the behavior when they switch over to ray 2.0? Is it possible to provide a tool to upgrade the yaml from the old to the new?

Actually, this is a good point -- our old default yaml have a bunch of **empty** deprecated fields.
We could have the code delete empty fields and log a warning.

**Comment by wuisawesome** [comment]
> We could have the code delete these empty fields and log a warning.

sgtm

File: python/ray/autoscaler/_private/util.py
Comment (DmitriGekhtman) [inline python/ray/autoscaler/_private/util.py:234]: Maybe something more prescriptive/informative like
`Dropping the empty legacy field {field}. {field} is not supported as of Ray 2.0.0. It is recommended to remove {field} from the cluster config.`

File: python/ray/autoscaler/_private/commands.py
Comment (wuisawesome) [reply python/ray/autoscaler/_private/commands.py:931]: agreed it's not perfect, but feels manageable to throw all the related hacks in one place here.

File: python/ray/tests/test_autoscaler.py
Comment (wuisawesome) [reply python/ray/tests/test_autoscaler.py:487]: it definitely took a hot sec to understand how to minimize changes here lol

**Comment by wuisawesome** [comment]
> Is it possible to provide a tool to upgrade the yaml from the old to the new?

We'll provide a guide on how to systematically do this, but it's not really practical to script it because they will need to fill out the `resources` section themselves.

**Comment by wuisawesome** [comment]
Migration docs are here: https://docs.google.com/document/d/1Rz-UGz-RHK6iKSX3xt2VioQnX-kC20d5n8pbRN7KGME/edit# 

I've done manual runs agains `aws/example-minimal.yaml` `aws/example-full.yaml` and `aws/example-full.yaml` (old version) and all behave as expected. Will merge once tests pass.

File: python/ray/tests/test_autoscaler.py
Comment (wuisawesome) [reply python/ray/tests/test_autoscaler.py:482]: oh dear god i definitely didn't run this again after making some changes...

File: python/ray/tests/test_autoscaler.py
Comment (DmitriGekhtman) [inline python/ray/tests/test_autoscaler.py:2614]: Whatever bug this is solving, it looks like it was painful to find.

**Comment by pcmoritz** [comment]
@wuisawesome Can you add the migration doc to the Ray documentation and link it in the error message users will be getting when they use an old autoscaler yaml?

**Pull request merged**
